### PR TITLE
Use EDU activity to determine user idleness

### PR DESCRIPTION
**Pull request by Half-Shot**
Whereas previously we've relied on users sending messages to prove if they are a lurker, this change would check to see if the users have sent any federated activity like presence, read receipts or typing. This allows the idle kicker to be much more fair about who it kicks.

### CODE REVIEW

File: src/bridge/IrcBridge.ts [hunk 1]
```diff
@@ -803,6 +808,34 @@ export class IrcBridge {
...
+        const event = request.getData();
+        let userIds: string[]|undefined = undefined;
+        if (!this.activityTracker) {
+            return;
+        }
+        if (event.type === "m.presence" && event.content.presence === "online") {
+            userIds = [event.sender];
+        }
+        else if (event.type === "m.receipt") {
+            userIds = Object.keys(Object.values(event.content).map((v) => v["m.read"]));
```
Comment (benparsons) [inline src/bridge/IrcBridge.ts:823]: Is this taking ALL the `m.read`s? Surely we still only want the sender in this point? Or you refresh the whole list whenever there's a new `m.receipt`?

**Review by benparsons — COMMENTED** [review]
question inline

File: src/bridge/IrcBridge.ts (hunk 1 above)
Comment (Half-Shot) [reply src/bridge/IrcBridge.ts:823]: Synapse sends new receipts in the format https://matrix.org/docs/spec/client_server/r0.6.0#m-receipt, which makes it a bit tricky to determine what's changed. We *could* do some quick filtering to exclude *stale* receipts though.

Comment (Half-Shot) [reply src/bridge/IrcBridge.ts]: I've tried something else.

**Review by benparsons — APPROVED** [review]
lgtm

**Pull request merged**